### PR TITLE
journalwatch: fix broken package, general maintenance

### PR DESCRIPTION
**Pull request by florianjacob**
###### Motivation for this change

The `journalwatch` package is broken on 19.03, as the now more strict dependency handling of `buildPythonPackage` uncovered the error of having `pytest` as `buildInput` instead of `checkInput`. This PR fixes that, along with some general maintenance:

* Move from deprecated `permissionsStartOnly`: https://github.com/NixOS/nixpkgs/pull/56265
* Move to `fetchFromGitHub`: https://github.com/NixOS/nixpkgs/issues/32997
* ~~Add `service.journalwatch.package` option for easy change of used package~~

The ee20ba8 fix pytest checks commit requires a backport to 19.03 to unbreak the package, the other commits are nice to have and should be backwards-compatible, but are not required. I'm sorry I did not notice the breakage prior release.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by florianjacob** [comment]
@GrahamcOfBorg build python36Packages.journalwatch

**Comment by florianjacob** [comment]
@GrahamcOfBorg build python36Packages.journalwatch

File: nixos/modules/services/logging/journalwatch.nix [hunk 1]
```diff
@@ -43,6 +56,15 @@ in {
         '';
       };
 
+      package = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/logging/journalwatch.nix]: I only see one version of this package in nixpkgs. Is this option necessary?

File: pkgs/tools/system/journalwatch/default.nix [hunk 2]
```diff
@@ -1,15 +1,16 @@
-{ stdenv, buildPythonPackage, fetchurl, pythonOlder, systemd, pytest }:
+{ stdenv, buildPythonPackage, fetchFromGitHub, pythonOlder, systemd, pytest }:
 
 buildPythonPackage rec {
   pname = "journalwatch";
   name = "${pname}-${version}";
```
Comment (aanderse) [inline pkgs/tools/system/journalwatch/default.nix]: While you're in here cleaning up would you mind removing this as it is no longer required?

File: nixos/modules/services/logging/journalwatch.nix [hunk 3]
```diff
@@ -199,33 +221,34 @@ in {
...
       isSystemUser = true;
-      createHome = true;
       home = dataDir;
-      # for journal access
-      group = "systemd-journal";
+      group = group;
     };
 
+    systemd.tmpfiles.rules = [
+      "d ${dataDir} 0750 ${user} ${group} - -"
```
Comment (aanderse) [inline nixos/modules/services/logging/journalwatch.nix]: Please replace with [StateDirectory and StateDirectoryMode](https://www.freedesktop.org/software/systemd/man/systemd.exec.html) as this is preferred over tmpfiles.d.

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/modules/services/logging/journalwatch.nix (hunk 1 above)
Comment (flokli) [reply nixos/modules/services/logging/journalwatch.nix]: IIRC, the way this should be done is via [overlays or packageOverrides](https://github.com/NixOS/nixpkgs/issues/55366#issuecomment-461780866) (the latter being [discouraged](https://github.com/NixOS/nixpkgs/pull/43560))

Comment (florianjacob) [reply nixos/modules/services/logging/journalwatch.nix]: I see; a working way to override the single pkgs.journalwatch with another version would indeed be enough and make this option unnecessary. Already did some research in that direction in the past, but never got it to work.

Could someone please give me a hand and tell me what I'm doing wrong here? All examples other than #55366 that I found only did add new packages or a full package set, but did not *replace* an existing package.

```nix
let
  mypkgsSource = builtins.fetchGit {
    # Descriptive name to make the store path easier to identify
    name = "florianjacob-nixpkgs";
    url = https://github.com/florianjacob/nixpkgs/;
    ref = "journalwatch";
    rev = "d5e303c9df8da2ae4aa42a7ccd28fbae4d24e30d";
  };
  mypkgs = import mypkgsSource {
    # does not help either
    # config = removeAttrs config.nixpkgs.config [ "packageOverrides" ];
  };
in
{
  # does not make a difference
  # nixpkgs.config.packageOverrides = pkgs: {
  #   # python3.pkgs.journalwatch = mypkgs.python3.pkgs.journalwatch;
  # };
  nixpkgs.overlays = [ (self: super: {
    python3.pkgs.journalwatch = mypkgs.python3.pkgs.journalwatch;
    # same error
    # python3Packages.journalwatch = mypkgs.python3Packages.journalwatch;
  }) ];

  # works fine:
  # services.journalwatch = {
  #   enable = true;
  #   package = mypkgs.python3Packages.journalwatch;
  # };
};
```

leads to

```
error: attribute 'buildPythonApplication' missing, at /nix/store/ai3gypqdhy8as9pik587cf7yx35546m2-nixos-19.03.172138.5c52b25283a/nixos/pkgs/development/tools/build-managers/meson/default.nix:3:1
```

File: pkgs/tools/system/journalwatch/default.nix (hunk 2 above)
Comment (florianjacob) [reply pkgs/tools/system/journalwatch/default.nix]: Good point, adressed.

File: nixos/modules/services/logging/journalwatch.nix (hunk 3 above)
Comment (florianjacob) [reply nixos/modules/services/logging/journalwatch.nix]: Also much more elegant, thank you! :+1:

File: nixos/modules/services/logging/journalwatch.nix (hunk 1 above)
Comment (aanderse) [reply nixos/modules/services/logging/journalwatch.nix]: @florianjacob sorry for taking a while to get back to you on this. Here is what I used:
```
  nixpkgsFJ = builtins.fetchGit {
    # Descriptive name to make the store path easier to identify
    name = "florianjacob-nixpkgs";
    url = https://github.com/florianjacob/nixpkgs/;
    ref = "journalwatch";
    rev = "60b9e339059e72d6ac3469509c41dacca57b2979";
  };
  pkgsFJ = import nixpkgsFJ { config = {}; };
```
...
```
  nixpkgs.overlays = [
    (self: super: {
      python3 = super.python3.override {
        packageOverrides = python-self: python-super: {
          journalwatch = pkgsFJ.python3Packages.journalwatch;
        };
      };
    })
  ];
```

Comment (florianjacob) [reply nixos/modules/services/logging/journalwatch.nix]: Works indeed, thanks alot! Removed the option now.

**Comment by florianjacob** [comment]
@aanderse Thanks for your review, it was very helpful! :beers:

**Comment by florianjacob** [comment]
@GrahamcOfBorg build python3Packages.journalwatch

**Review by aanderse — APPROVED** [review]

**Comment by aanderse** [comment]
@Ekleog I'm a big fan of your `system-sendmail' package. How do you feel about https://github.com/NixOS/nixpkgs/commit/c29abbcfcbca7b5b88dfb0e49706ef12f4800c80?

**Comment by flokli** [comment]
@aanderse `system-sendmail` looks nice - didn't know it so far.
I guess it makes sense to use it in journalwatch, too :-)

Maybe we want to change other places referencing `/run/wrappers/bin/sendmail` directly too? (in a followup PR)

**Comment by aanderse** [comment]
@flokli that sounds good. Hopefully at some point I'll circle back to that.
@florianjacob awesome job!

**Comment by florianjacob** [comment]
Forgot the backporting, PR at #59424.

**Pull request merged**